File: core/memory/runtime_merge_engine.py
from __future__ import annotations

from typing import Any, Dict, List

from core.memory.runtime_federation_engine import federate_runtime_memory
from core.memory.runtime_memory_engine import build_runtime_memory


def merge_runtime_memories(
    memories: List[Dict[str, Any]],
) -> Dict[str, Any]:
    ordered = sorted(
        memories,
        key=lambda m: str(m.get("memory_id", m.get("runtime_id", ""))),
    )
    for mem in ordered:
        history = mem.get("runtime_history", [])
        if isinstance(history, list):
            mem["runtime_history"] = sorted(
                history,
                key=lambda h: (
                    int(h.get("tick", 0)),
                    str(h.get("kind", "")),
                    str(h.get("source", "")),
                ),
            )
    federated = federate_runtime_memory(ordered)
    return build_runtime_memory(
        runtime_history=federated.get("runtime_history", []),
        lineage=federated.get("lineage", []),
        semantic_relations=federated.get("semantic_relations", []),
    )

File: src/memory/runtimeMergeEngine.ts
/**
 * Converted from Python: core/memory/runtime_merge_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { federateRuntimeMemory } from "./runtimeFederationEngine.js";
import { buildRuntimeMemory } from "./runtimeMemoryEngine.js";

export function mergeRuntimeMemories(memories: any): any {
  var ordered: any = py.sorted(memories, {key: ((m: any) => py.toStr(py.get(m, "memory_id", py.get(m, "runtime_id", "")))) as (item: any) => any});
  var mem: any;
  for (mem of py.iter(ordered)) {
    var history: any = py.get(mem, "runtime_history", []);
    if ((Array.isArray(history))) {
      py.setItem(mem, "runtime_history", py.sorted(history, {key: ((h: any) => [py.toInt(py.get(h, "tick", 0)), py.toStr(py.get(h, "kind", "")), py.toStr(py.get(h, "source", ""))]) as (item: any) => any}));
    }
  }
  var federated: any = federateRuntimeMemory(ordered);
  return buildRuntimeMemory(py.get(federated, "runtime_history", []), py.get(federated, "lineage", []), py.get(federated, "semantic_relations", []));
}
export { buildRuntimeMemory, federateRuntimeMemory };
